### PR TITLE
[SPARK-31220][SQL] repartition obeys initialPartitionNum when adaptiveExecutionEnabled

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?
This PR makes `repartition`/`DISTRIBUTE BY` obeys [initialPartitionNum](https://github.com/apache/spark/blob/af4248b2d661d04fec89b37857a47713246d9465/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L446-L455) when adaptive execution enabled.

### Why are the changes needed?
To make `DISTRIBUTE BY`/`GROUP BY` partitioned by same partition number.
How to reproduce:
```scala
spark.sql("CREATE TABLE spark_31220(id int)")
spark.sql("set spark.sql.adaptive.enabled=true")
spark.sql("set spark.sql.adaptive.coalescePartitions.initialPartitionNum=1000")
```

Before this PR:
```
scala> spark.sql("SELECT id from spark_31220 GROUP BY id").explain
== Physical Plan ==
AdaptiveSparkPlan(isFinalPlan=false)
+- HashAggregate(keys=[id#5], functions=[])
   +- Exchange hashpartitioning(id#5, 1000), true, [id=#171]
      +- HashAggregate(keys=[id#5], functions=[])
         +- FileScan parquet default.spark_31220[id#5]

scala> spark.sql("SELECT id from spark_31220 DISTRIBUTE BY id").explain
== Physical Plan ==
AdaptiveSparkPlan(isFinalPlan=false)
+- Exchange hashpartitioning(id#5, 200), false, [id=#179]
   +- FileScan parquet default.spark_31220[id#5]
```
After this PR:
```
scala> spark.sql("SELECT id from spark_31220 GROUP BY id").explain
== Physical Plan ==
AdaptiveSparkPlan(isFinalPlan=false)
+- HashAggregate(keys=[id#5], functions=[])
   +- Exchange hashpartitioning(id#5, 1000), true, [id=#171]
      +- HashAggregate(keys=[id#5], functions=[])
         +- FileScan parquet default.spark_31220[id#5]

scala> spark.sql("SELECT id from spark_31220 DISTRIBUTE BY id").explain
== Physical Plan ==
AdaptiveSparkPlan(isFinalPlan=false)
+- Exchange hashpartitioning(id#5, 1000), false, [id=#179]
   +- FileScan parquet default.spark_31220[id#5]
```

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

**Comment by wangyum** [comment]
cc @cloud-fan @HyukjinKwon @dongjoon-hyun @maryannxue Do we need this change to make `DISTRIBUTE BY`/`GROUP BY` partitioned by same partition number?

**Comment by HyukjinKwon** [comment]
From a cursory look, I think it makes sense.

**Comment by koertkuipers** [comment]
adaptive execution estimates the number of partitions for a shuffle using `spark.sql.adaptive.shuffle.targetPostShuffleInputSize` as its target size per shuffled partition. i was surprised to find out however that it does not do this for a `DataFrame.repartition(...)`. i dont understand why since under the hood its also just a shuffle no different than a `DataFrame.groupBy`.
will this pull request fix this issue? from looking at code i dont understand if it does so, it doesnt look like it to me.

**Comment by wangyum** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -1021,4 +1021,20 @@ class AdaptiveQueryExecSuite
       }
     }
   }
+
+  test("SPARK-31220 repartition obeys initialPartitionNum when adaptiveExecutionEnabled") {
+    Seq(true, false).foreach { adaptiveExecutionEnabled =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1026]: can we make this shorter? `enableAQE`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 2]
```diff
@@ -1021,4 +1021,20 @@ class AdaptiveQueryExecSuite
...
+      withSQLConf(
+        SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> s"$adaptiveExecutionEnabled",
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1028]: `enableAQE.toString`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: Fixed it.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by koertkuipers** [comment]
i rebuild spark using this pullreq hoping to have `def repartition(partitionExprs: Column*)` now use adaptive execution to scale the number of reducers. but i see no such behavior.
my simple test job only does 3 things:
1. read from parquet datasource
2. repartition by one column
3. write to parquet datasink

what i see is that the actual number of shuffle partitions is always equal to `spark.sql.shuffle.partitions` or to `spark.sql.adaptive.coalescePartitions.initialPartitionNum` if it has been set. but it never adapts to the data size as i would expect with adapative execution (which it does with a groupBy).

**Comment by wangyum** [comment]
@koertkuipers The shuffle write step is always `spark.sql.adaptive.coalescePartitions.initialPartitionNum`. 
Shuffle read step will coalesce if `spark.sql.adaptive.coalescePartitions.enabled=true` and the data size is small.

**Comment by koertkuipers** [comment]
@wangyum i have `spark.sql.adaptive.coalescePartitions.enabled=true` and the data size is small.
how can i see that the step does coalesce? in number of tasks (i always see 2048)? in number of output files (i always see 2048)?

i have `spark.sql.shuffle.partitions=2048` and `spark.sql.adaptive.coalescePartitions.initialPartitionNum=2048`.

when i do a groupBy instead of repartition than the number of tasks varies with data size (and is less than 2048) and the number of output files varies too (and is less than 2048). with repartition the tasks and output files are always fixed at 2048.

**Comment by HyukjinKwon** [comment]
@koertkuipers, are you able to show the exact reproducible steps? Sounds like ideally we should file a separate JIRA.

**Comment by cloud-fan** [comment]
note: `repartition` is a physical operation that users specify how many partitions to have after the shuffle. That's why Spark can't coalesce the partition.

Other optimization can still happen, like SMJ to BHJ.

**Comment by koertkuipers** [comment]
@cloud-fan so how can i repartition by a column without having to specify the number of partitions?

**Comment by cloud-fan** [comment]
repartition by key is another story and should support partition coalesce. If it's not the case now, please create a bug report in the JIRA, thanks!

**Comment by koertkuipers** [comment]
@cloud-fan ok i will create jira for repartition by key with partition coalesce for AQE

**Comment by koertkuipers** [comment]
@cloud-fan see https://issues.apache.org/jira/browse/SPARK-32056

**Pull request closed without merging**